### PR TITLE
[natspec] Introduce AST node for structured documentation

### PR DESCRIPTION
**Pull request by erak**
Part of https://github.com/ethereum/solidity/issues/7835.

The idea behind this PR is to introduce a new AST node which is added as a subnode for all AST nodes that can be documented using a formal specification, here NatSpec (contract, function, event and modifier definitions). For all other nodes the documentation is still a plain string.

### CODE REVIEW

File: libsolidity/ast/AST_accept.h [hunk 1]
```diff
@@ -69,6 +69,18 @@ void ImportDirective::accept(ASTConstVisitor& _visitor) const
 	_visitor.endVisit(*this);
 }
 
+void DocString::accept(ASTVisitor& _visitor)
```
Comment (chriseth) [inline libsolidity/ast/AST_accept.h]: Const visitor is missing

File: libsolidity/analysis/DocStringAnalyser.cpp [hunk 2]
```diff
@@ -123,15 +123,29 @@ void DocStringAnalyser::parseDocStrings(
...
+				{
+					bool foundMatch = false;
+					for (auto const& returnParameter: function->returnParameters())
+					{
+						auto firstWord = docTag.second.content.substr(docTag.second.content.find_first_of(" \t"));
+						if (firstWord == returnParameter->name())
+							foundMatch = true;
+					}
+					if (!foundMatch)
+						appendError("Doc tag @" + docTag.first + " does not match named return parameter.");
```
Comment (chriseth) [inline libsolidity/analysis/DocStringAnalyser.cpp]: Do we call it "Doc tag" in other places?

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -268,7 +268,7 @@ bool ASTJsonConverter::visit(ContractDefinition const& _node)
 {
 	setJsonNode(_node, "ContractDefinition", {
 		make_pair("name", _node.name()),
-		make_pair("documentation", _node.documentation() ? Json::Value(*_node.documentation()) : Json::nullValue),
+		make_pair("documentation", _node.documentation()->text() ? Json::Value(*_node.documentation()->text()) : Json::nullValue),
```
Comment (erak) [inline libsolidity/ast/ASTJsonConverter.cpp]: I think we should append an AST node to the JSON output here. So this should not be the final solution.

File: libsolidity/analysis/DocStringAnalyser.cpp (hunk 2 above)
Comment (erak) [reply libsolidity/analysis/DocStringAnalyser.cpp]: That particular change is not part of this PR any more. Moved it to https://github.com/ethereum/solidity/pull/7837.
But yes, we do use that wording. Please see the other error reported in the very same function.

File: libsolidity/parsing/Parser.cpp [hunk 4]
```diff
@@ -135,6 +135,17 @@ void Parser::parsePragmaVersion(SourceLocation const& _location, vector<Token> c
 			);
 }
 
+ASTPointer<NatspecDocumentation> Parser::parseNatspecDocumentation()
+{
+	ASTNodeFactory nodeFactory(*this);
+	// TODO The line below leads to an ICE because the source is null
+	// at that location.
+	// nodeFactory.setLocation(m_scanner->currentCommentLocation());
```
Comment (erak) [inline libsolidity/parsing/Parser.cpp]: This needs to be addressed.

File: test/libsolidity/ASTJSON/documentation.json [hunk 5]
```diff
@@ -16,17 +16,23 @@
       "baseContracts": [],
       "contractDependencies": [],
       "contractKind": "contract",
-      "documentation": "This contract is empty",
+      "documentation":
+      {
+        "id": 1,
+        "nodeType": "Natspec",
+        "src": "28:8:1",
```
Comment (erak) [inline test/libsolidity/ASTJSON/documentation.json]: Source location is still incorrect.

File: libsolidity/ast/AST.h [hunk 6]
```diff
@@ -362,6 +386,24 @@ class Documented
 	ASTPointer<ASTString> m_documentation;
 };
 
+/**
+ * Abstract class that is added to each AST node that can receive documentation.
+ */
+class NatspecDocumented
```
Comment (Marenz) [inline libsolidity/ast/AST.h]: Introducing this class means you also have to update all `accept` functions of all impacted classes to visit the `m_documentation` member, right?

File: liblangutil/Scanner.cpp [hunk 7]
```diff
@@ -429,6 +429,7 @@ Token Scanner::scanSlash()
 			Token comment;
 			m_skippedComments[NextNext].location.start = firstSlashPosition;
 			comment = scanSingleLineDocComment();
+			m_skippedComments[NextNext].location.source = m_source;
```
Comment (erak) [inline liblangutil/Scanner.cpp:431]: The node factory requires a valid source for a location when creating a new node. Also needed for proper error reporting later on.

File: libsolidity/ast/ASTAnnotations.h [hunk 8]
```diff
@@ -142,7 +142,7 @@ struct VariableDeclarationAnnotation: DeclarationAnnotation
 	std::set<CallableDeclaration const*> baseFunctions;
 };
 
-struct StatementAnnotation: ASTAnnotation, DocumentedAnnotation
+struct StatementAnnotation: ASTAnnotation
```
Comment (erak) [inline libsolidity/ast/ASTAnnotations.h:145]: This is a rather unrelated change. But right now, we do not access the documentation annotations of statements anywhere. Therefore it was removed from the base class list

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 9]
```diff
@@ -833,6 +833,18 @@ bool ASTJsonConverter::visit(Literal const& _node)
 	return false;
 }
 
+bool ASTJsonConverter::visit(FormalDocumentation const& _node)
+{
+	Json::Value text{*_node.text()};
+	std::vector<pair<string, Json::Value>> attributes = {
+		make_pair("text", text),
+		make_pair("format", "NatSpec")
```
Comment (erak) [inline libsolidity/ast/ASTJsonConverter.cpp]: This indicates that the NatSpec format used to parse documentation in source. Since it's the only available format right now, it's hard-coded here and in the AST JSON importer. Could be replaced by an enum in `FormalDocumentation` in the future if needed.

File: libsolidity/ast/AST.h [hunk 10]
```diff
@@ -345,6 +345,32 @@ class VariableScope
 	std::vector<VariableDeclaration const*> m_localVariables;
 };
 
+/**
+ * The formal documentation class that represents an AST node, which is created for
+ * other AST nodes that can receive such.
+ * Currently, NatSpec is the only supported specification.
+ */
+class FormalDocumentation: public ASTNode
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: Maybe rather `StructuredDocumentation`?

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 11]
```diff
@@ -842,6 +844,23 @@ ASTPointer<ASTNode> ASTJsonImporter::createLiteral(Json::Value const&  _node)
...
 }
 
+ASTPointer<FormalDocumentation> ASTJsonImporter::createDocumentation(Json::Value const&  _node)
+{
+	static string const textString = "text";
+	static string const formatString = "format";
+
+	astAssert(member(_node, textString).isString(), "'text' must be a string");
+	astAssert(member(_node, formatString).isString(), "'format' must be a string");
+	astAssert(member(_node, formatString) == "NatSpec", "'format' must specify 'NatSpec'");
```
Comment (erak) [inline libsolidity/ast/ASTJsonImporter.cpp]: I've added a comment about the hard-coded format check here: https://github.com/ethereum/solidity/pull/7834/files#r372672532.

File: libsolidity/parsing/Parser.cpp [hunk 12]
```diff
@@ -135,6 +135,15 @@ void Parser::parsePragmaVersion(SourceLocation const& _location, vector<Token> c
 			);
 }
 
+ASTPointer<FormalDocumentation> Parser::parseFormalDocumentation()
+{
+	ASTNodeFactory nodeFactory{*this};
+	nodeFactory.setLocation(m_scanner->currentCommentLocation());
```
Comment (erak) [inline libsolidity/parsing/Parser.cpp]: Here the location is set manually, and therefore https://github.com/ethereum/solidity/pull/7834/files#r372670260 is needed.

File: libsolidity/ast/ASTAnnotations.h (hunk 8 above)
Comment (chriseth) [reply libsolidity/ast/ASTAnnotations.h:145]: Yes, this is a remnant of the why3 mechanism for formal verification.

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 9 above)
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: Why specify this?

File: libsolidity/ast/AST_accept.h [hunk 13]
```diff
@@ -67,6 +67,18 @@ void ImportDirective::accept(ASTConstVisitor& _visitor) const
 	_visitor.endVisit(*this);
 }
```
Comment (chriseth) [inline libsolidity/ast/AST_accept.h:69]: This needs changes in all ast nodes that inherit from 'FormallyDocumented'

File: libsolidity/parsing/Parser.cpp [hunk 14]
```diff
@@ -276,14 +285,14 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
...
 	ASTNodeFactory nodeFactory(*this);
 	ASTPointer<ASTString> name =  nullptr;
-	ASTPointer<ASTString> docString;
+	ASTPointer<FormalDocumentation> documentation;
 	vector<ASTPointer<InheritanceSpecifier>> baseContracts;
 	vector<ASTPointer<ASTNode>> subNodes;
 	std::pair<ContractKind, bool> contractKind{};
 	try
 	{
 		if (m_scanner->currentCommentLiteral() != "")
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Can you pull this condition into `parseFormalDocumentation`?

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 9 above)
Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: My issue is that this is not really natspec. This is just doxygen / javadoc. Natspec is the formatting inside of `@userdoc`.

Comment (erak) [reply libsolidity/ast/ASTJsonConverter.cpp]: I had the impression that it would be better to be able to switch the formal / structured doc format. At least under the assumption that NatSpec is not accepted thoughout the community. It's not needed right now, but I wanted to take the chance within this change to support it. Do you think it's not necessary and could be added later?

File: libsolidity/ast/AST_accept.h (hunk 13 above)
Comment (erak) [reply libsolidity/ast/AST_accept.h:69]: So that would be
```
if (m_documentation)
    m_documentation->accept(_visitor);
```
Or am I missing something else?

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 9 above)
Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: Especially since the ast json output is solidity version specific, I don't think we need the ability to switch.

File: libsolidity/ast/AST_accept.h (hunk 13 above)
Comment (chriseth) [reply libsolidity/ast/AST_accept.h:69]: Yes!

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 9 above)
Comment (erak) [reply libsolidity/ast/ASTJsonConverter.cpp]: Clarified in the chat. I'll remove the attribute then. Thanks.

File: libsolidity/ast/ASTJsonImporter.cpp (hunk 11 above)
Comment (erak) [reply libsolidity/ast/ASTJsonImporter.cpp]: Removed again.

File: libsolidity/ast/AST.h (hunk 10 above)
Comment (erak) [reply libsolidity/ast/AST.h]: Sounds better. Changed it accordingly.

File: libsolidity/ast/AST.h [hunk 15]
```diff
@@ -345,6 +345,31 @@ class VariableScope
...
+	{}
+
+	void accept(ASTVisitor& _visitor) override;
+	void accept(ASTConstVisitor& _visitor) const override;
+
+	ASTPointer<ASTString> const& text() const { return m_text; }
+
+private:
+
+	///
```
Comment (erak) [inline libsolidity/ast/AST.h]: Ah, this is missing a comment.

File: libsolidity/ast/AST.h (hunk 6 above)
Comment (erak) [reply libsolidity/ast/AST.h]: Done.

File: Changelog.md [hunk 16]
```diff
@@ -1,7 +1,7 @@
 ### 0.6.3 (unreleased)
 
 Language Features:
-
+ * AST: Add a new node for doxygen-style, structured documentation that can be received by contract, function, event and modifier definitions.
```
Comment (chriseth) [inline Changelog.md]: I would say this is a compiler feature.

Comment (chriseth) [reply Changelog.md]: Adding it to VariableDeclaration is a next step?

File: liblangutil/Scanner.cpp (hunk 7 above)
Comment (chriseth) [reply liblangutil/Scanner.cpp:431]: Why not assign it rigth after start (all others need `scanSingleLineDocComment` to run first)? The same below.

File: libsolidity/ast/AST.h [hunk 17]
```diff
@@ -345,6 +345,31 @@ class VariableScope
 	std::vector<VariableDeclaration const*> m_localVariables;
 };
 
+/**
+ * The doxygen-style, structured documentation class that represents an AST node,
+ * which is created for other AST nodes that can receive such.
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: This line is redundant because it is valid for all AST nodes apart from the topmost one :)

File: libsolidity/ast/AST.h [hunk 18]
```diff
@@ -345,6 +345,31 @@ class VariableScope
 	std::vector<VariableDeclaration const*> m_localVariables;
 };
 
+/**
+ * The doxygen-style, structured documentation class that represents an AST node,
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: Maybe comment that the "structured documentation" is just stored as a single string and not further split up as part of the AST.

Comment (chriseth) [reply libsolidity/ast/AST.h]: Or will that change soon?

**Review by Marenz — DISMISSED** [review]

File: libsolidity/ast/Types.cpp [hunk 19]
```diff
@@ -3323,13 +3323,13 @@ Type const* FunctionType::selfType() const
 	return m_parameterTypes.at(0);
 }
 
-ASTPointer<ASTString> FunctionType::documentation() const
+ASTPointer<StructuredDocumentation> FunctionType::documentation() const
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp:3326]: We should maybe change this later so that it returns a plain pointer to StructuredDocumentation. I don't see why we should create a new shared_ptr below in the failure case.

**Comment by chriseth** [comment]
Looks good apart from the small changes to the comments!

File: libsolidity/ast/AST.h (hunk 18 above)
Comment (erak) [reply libsolidity/ast/AST.h]: That will change soon, but makes sense for now :)

File: libsolidity/ast/AST.h (hunk 17 above)
Comment (erak) [reply libsolidity/ast/AST.h]: ok!

File: Changelog.md (hunk 16 above)
Comment (erak) [reply Changelog.md]: Makes sense.

Comment (erak) [reply Changelog.md]: > Adding it to VariableDeclaration is a next step?

I would say yes, because that's a low hanging fruit. The step after would be to properly report source locations (code is there, PR will follow)

File: liblangutil/Scanner.cpp (hunk 7 above)
Comment (erak) [reply liblangutil/Scanner.cpp:431]: Done.

File: libsolidity/ast/AST.h (hunk 18 above)
Comment (erak) [reply libsolidity/ast/AST.h]: Maybe I'll not add that comment, since it's changed very soon anyways.

**Review by chriseth — APPROVED** [review]
